### ISSUE TITLE
Incorrect error message when date format is wrong

### ISSUE THREAD

**Issue by baptistemarchand** [prisma/prisma#20959]
### Bug description

When filtering on a date and using a wrong date format (e.g: `2023-01-01` instead of the full ISO string), we sometimes get incorrect error messages that point us in a different direction, which is very confusing.
It only happens when the field we are filtering on is in a relation.

First, here is an example of Prisma working as intended (`departureTime` is a `DateTime` field) :
```typescript
await prisma.trip.count({
  where: {
    assignedVehicle: {
      not: null,
    },
    departureTime: {
      gte: 'NOT A DATE',
    },
  },
})
```
We get the following error message which is good :
```
Argument departureTime: Got invalid value
{
  gte: 'NOT A DATE'
}
on prisma.aggregateTrip. Provided Json, expected DateTimeNullableFilter or DateTime or Null.
```

But now if we change the code slightly to filter on a relation :
```typescript
await prisma.passenger.count({
    where: {
      trip: {
        assignedVehicle: {
          not: null,
        },
        departureTime: {
          gte: 'NOT A DATE',
        },
      },
    },
  })
```
We get this error :
```
Unknown arg `assignedVehicle` in where.trip.assignedVehicle for type TripRelationFilter. Available args:

type TripRelationFilter {
  is?: TripWhereInput | Null
  isNot?: TripWhereInput | Null
}
Unknown arg `departureTime` in where.trip.departureTime for type TripRelationFilter. Available args:

type TripRelationFilter {
  is?: TripWhereInput | Null
  isNot?: TripWhereInput | Null
}
```
This error message is very confusing because it does not mention that something is wrong with the date, and even talks about the field `assignedVehicle` even though there is nothing wrong with it.

### How to reproduce

See code above, and prisma schema below.

### Expected behavior

When I run this :
```typescript
await prisma.passenger.count({
    where: {
      trip: {
        assignedVehicle: {
          not: null,
        },
        departureTime: {
          gte: 'NOT A DATE',
        },
      },
    },
  })
```
I expected to get a similar error as in my first example.

### Prisma information

```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["extendedWhereUnique"]
}

datasource db {
  provider = "mysql"
  url      = env("DATABASE_URL")
}

model Trip {
  id        String   @id @default(cuid()) @db.Char(25)
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  departureTime       DateTime?
  assignedVehicle    String
  passengers Passenger[]
}

model Passenger {
  id              String   @id @default(cuid()) @db.Char(25)
  createdAt       DateTime @default(now())
  updatedAt       DateTime @updatedAt
  tripId          String   @map("trip") @db.Char(25)
  trip            Trip     @relation(fields: [tripId], references: [id], onUpdate: Restrict, map: "Passenger_ibfk_2")

  @@index([tripId], map: "trip")
}

```

### Environment & setup

- OS: macOS
- Database: MySQL
- Node.js version: v16.14.0

### Prisma Version

```
4.10.0
```

**Comment by aqrln** [comment]
> This error message is very confusing because it does not mention that something is wrong with the date, and even talks about the field assignedVehicle even though there is nothing wrong with it.

There is, in fact, a problem with `assignedVehicle`: it's not nullable in the schema so `not: null` filter is not valid as it can never be `null`.

I agree that it is not immediately clear what exactly is wrong with this field from the error message in Prisma 4.10.0, but if you apply the suggestion from this error and use an explicit `is` instead of the shortcut, the error becomes more understandable. That said, with latest Prisma 5 there's no need to do even that, as even when using the shortcuts the error message is very clear and specific:
```
PrismaClientValidationError:
Invalid `prisma.passenger.count()` invocation in
/Users/aqrln/prisma/prisma/reproductions/current/index.ts:23:26

  20
  21 async function main() {
  22   const prisma = new PrismaClient()
→ 23   await prisma.passenger.count({
         select: {
           _count: {
             select: {
               _all: true
             }
           }
         },
         where: {
           trip: {
             assignedVehicle: {
       +       not: String
             },
             departureTime: {
               gte: "NOT A DATE"
             }
           }
         }
       })

Argument `not` must not be null.
```

The static type error, unlike the runtime validation error, is unfortunately a bit confusing though.

After fixing that error, this query
```ts
  await prisma.passenger.count({
    where: {
      trip: {
        departureTime: {
          gte: 'NOT A DATE'
        }
      },
    },
  })
```
produces the following validation error:
```
Invalid `prisma.passenger.count()` invocation in
/Users/aqrln/prisma/prisma/reproductions/current/index.ts:23:26

  20
  21 async function main() {
  22   const prisma = new PrismaClient()
→ 23   await prisma.passenger.count({
         select: {
           _count: {
             select: {
               _all: true
             }
           }
         },
         where: {
           trip: {
             departureTime: {
               gte: "NOT A DATE"
                    ~~~~~~~~~~~~
             }
           }
         }
       })

Invalid value for argument `gte`: input contains invalid characters. Expected ISO-8601 DateTime.
```

I'll go ahead and close the issue since date format validation is working as expected, and the validation error about the other issue hopefully shouldn't have led to any confusion when using recent versions of Prisma. Besides the new validation engine, JSON protocol that has become the default in Prisma 5 also brings significant performance improvements, so I'd recommend upgrading.

**Comment by baptistemarchand** [comment]
Thank you for your detailed and fast answer. I'll try to upgrade.

P.S: I made a mistake when I copied my Prisma schema in the Github issue. In my real schema `assignedVehicle` is nullable. It was an enum so I changed it to a String to make the example easier to reproduce but forgot to add the nullable part. So I still think there was problem there. But it doesn't matter if it's fixed in a more recent version!
